File: extensions/intellij/src/main/java/com/voidmuse/idea/plugin/mcp/MCPClient.java
package com.voidmuse.idea.plugin.mcp;

import io.modelcontextprotocol.client.McpClient;
import io.modelcontextprotocol.client.McpSyncClient;
import io.modelcontextprotocol.client.transport.HttpClientSseClientTransport;
import io.modelcontextprotocol.client.transport.HttpClientStreamableHttpTransport;
import io.modelcontextprotocol.client.transport.ServerParameters;
import io.modelcontextprotocol.client.transport.StdioClientTransport;
import io.modelcontextprotocol.spec.McpClientTransport;
import io.modelcontextprotocol.spec.McpSchema;

import java.net.http.HttpRequest;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.logging.Logger;

public class MCPClient {
    private static final Logger LOGGER = Logger.getLogger(MCPClient.class.getName());

    private McpSyncClient client;
    private List<McpSchema.Tool> tools = new ArrayList<>();
    private McpItem mcpConfig;

    public MCPClient(McpItem mcpConfig) {
        this.mcpConfig = mcpConfig;
    }

    public void init(){
        McpClientTransport transport;
        LOGGER.info("Initializing MCP client for: " + mcpConfig.getName());
        LOGGER.info("Command: " + mcpConfig.getCommand());
        LOGGER.info("Args: " + mcpConfig.getArgs());
        LOGGER.info("URL: " + mcpConfig.getUrl());
        LOGGER.info("Timeout: 30 seconds");
        
        if (mcpConfig.getUrl() != null && !mcpConfig.getUrl().isEmpty()) {

            try {
                HttpClientSseClientTransport.Builder builder = HttpClientSseClientTransport.builder(mcpConfig.getUrl());
                if(mcpConfig.getHeaders() != null && !mcpConfig.getHeaders().isEmpty()){
                    Consumer<HttpRequest.Builder> customizer = reqBuilder -> {
                        mcpConfig.getHeaders().forEach((k,v)->{
                            reqBuilder.setHeader(k,v.toString());
                        });
                    };

                    builder.customizeRequest(customizer);
                }
                transport = builder.build();

                this.client = McpClient.sync(transport)
                        .requestTimeout(Duration.ofSeconds(30))
                        .capabilities(McpSchema.ClientCapabilities.builder()
                                .roots(true)      // Enable roots capability
                                .sampling()       // Enable sampling capability
                                .build())
                        .build();
                this.client.initialize();

                LOGGER.info("Connected using SSE transport");
            } catch (Exception e) {
                LOGGER.severe("Failed to initialize HTTP transport: " + e.getMessage());
                e.printStackTrace();
                throw new RuntimeException("Failed to initialize MCP client: " + e.getMessage(), e);
            }
        } else{
            try {
                LOGGER.info("Using stdio transport with command: " + mcpConfig.getCommand());
                ServerParameters params = ServerParameters.builder(mcpConfig.getCommand())
                        .args(mcpConfig.getArgs())
                        .build();
                transport = new StdioClientTransport(params);
                this.client = McpClient.sync(transport)
                        .requestTimeout(Duration.ofSeconds(30))
                        .capabilities(McpSchema.ClientCapabilities.builder()
                                .roots(true)      // Enable roots capability
                                .sampling()       // Enable sampling capability
                                .build())
                        .build();
                LOGGER.info("Initializing stdio client...");
                this.client.initialize();
                LOGGER.info("Connected using stdio transport");
            } catch (Exception e) {
                LOGGER.severe("Failed to initialize stdio transport: " + e.getMessage());
                e.printStackTrace();
                throw new RuntimeException("Failed to initialize MCP client: " + e.getMessage(), e);
            }
        }
        try {
            McpSchema.ListToolsResult tools = client.listTools();
            this.tools = tools.tools();
            mcpConfig.setConnected(true);
            LOGGER.info("Successfully loaded " + tools.tools().size() + " tools");
        } catch (Exception e) {
            LOGGER.severe("Failed to list tools: " + e.getMessage());
            e.printStackTrace();
            throw new RuntimeException("Failed to list tools: " + e.getMessage(), e);
        }
    }

    public void close() {
        if (client != null) {
            client.close();
        }
    }

    public List<McpSchema.Tool> getTools() {
        return tools;
    }

    public McpItem getConfig() {
        return mcpConfig;
    }

    public Object callTool(String name, Map<String, Object> params) throws Exception {
        McpSchema.CallToolRequest request = new McpSchema.CallToolRequest(name,params);
        return client.callTool(request);
    }

}

File: extensions/vscode/src/services/McpClient.ts
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { Tool } from "@modelcontextprotocol/sdk/types.js";

export interface McpItem {
    key: string;
    name: string;
    url?: string;
    command?: string;
    args?: string[];
    headers?: Record<string, string>;
    connected?: boolean;
    enabled?: boolean;
    tools?: {
        name: string;
        description: string | undefined; 
    }[];
}

export class McpClient {
    private client: Client;
    private tools: Tool[] = [];
    public mcpConfig: McpItem;

    constructor(mcpConfig: McpItem) {
        this.client = new Client({
                name: 'voidmuse-mcp-client',
                version: '1.0.0'
            });
        this.mcpConfig = mcpConfig;
    }

    public async init() {
        await this.connectToServer();
    }

    public async close() {
        await this.client.close();
    }

    public async getTools() {
        return this.tools;
    }

    public getConfig() {
        return this.mcpConfig;
    }

    public async callTool(name: string, params: Record<string, any>) {
        return await this.client.callTool({
            name,
            arguments: params,
        });
    }

    private async connectToServer() {
        //连接远程mcp服务
        if(this.mcpConfig.url){
            const baseUrl = new URL(this.mcpConfig.url);
            try {
                this.client = new Client({
                    name: 'streamable-http-client',
                    version: '1.0.0'
                });
                let transport = new StreamableHTTPClientTransport(
                    new URL(baseUrl)
                );
                if (this.mcpConfig.headers){
                    transport = new StreamableHTTPClientTransport(baseUrl,{
                    requestInit: { headers: this.mcpConfig.headers }
                    });
                }
                await this.client.connect(transport);
                console.log("Connected using Streamable HTTP transport");
            } catch (error) {
                // If that fails with a 4xx error, try the older SSE transport
                console.log("Streamable HTTP connection failed, falling back to SSE transport");
                this.client = new Client({
                    name: 'sse-client',
                    version: '1.0.0'
                });
                let sseTransport = new SSEClientTransport(baseUrl);
                if (this.mcpConfig.headers){
                    sseTransport = new SSEClientTransport(baseUrl,{
                    requestInit: { headers: this.mcpConfig.headers }
                    });
                }
        

                await this.client.connect(sseTransport);
                console.log("Connected using SSE transport");
            }
        }
        //连接本地mcp服务
        else if(this.mcpConfig.command){
            const stdioTransport = new StdioClientTransport({
                command: this.mcpConfig.command,
                args: this.mcpConfig.args,
                env: {}
            });
            this.client.connect(stdioTransport);
            console.log("Connected using stdio transport");
        }
        
        const toolsResult = this.client.listTools();

        const tools = (await toolsResult).tools;
        console.log(tools);
        this.tools = tools.map((tool) => {
            return {
                name: tool.name,
                description: tool.description,
                inputSchema: tool.inputSchema,
            };
        });
        this.mcpConfig.connected = true;

        console.log(
            "Connected to server with tools:",
            this.tools.map(({ name }) => name)
        );
    }
}
